Tasks/DeployAzureResourceGroup/MachineGroupAgentExtensionManager.ts

```diff
@@ -0,0 +1,179 @@
+            }
+            return this.deferred.promise;
+        }
+        catch (exception) {
+            this.log(tl.loc("MGAgentOperationOnAllVMsFailed", exception.message));
+            tl.setResult(tl.TaskResult.Failed, tl.loc("MGAgentOperationOnAllVMsFailed", exception.message));
+            this.deferred.resolve(tl.loc("MGAgentOperationOnAllVMsFailed", exception.message));
+        }
+    }
+
+    public async deleteMGExtension() {
+        try {
+            var operation = "uninstallation";
+            this.deferred = Q.defer<string>();
+            var listOfVms = await this.azureUtils.getVMDetails();
+            this.vmCount = listOfVms.length;
+            if (this.vmCount == 0) {
+                this.deferred.resolve("");
+            }
+            for (var i = 0; i < listOfVms.length; i++) {
+                var vmName = listOfVms[i]["name"];
+                var resourceGroupName = this.taskParameters.resourceGroupName;
+                var extensionParameters = this.formExtensionParameters(listOfVms[i], operation);
+                this.log(tl.loc("DeleteExtension", extensionParameters["extensionName"], vmName));
+                var extensionName = extensionParameters["extensionName"];
+                var callback = this.givePostOperationCallBack(extensionName, vmName, operation);
+                this.computeClient.virtualMachineExtensions.deleteMethod(resourceGroupName, vmName, extensionName, callback);
+            }
+            return this.deferred.promise;
+        }
+        catch (exception) {
+            this.log(tl.loc("MGAgentOperationOnAllVMsFailed", operation, exception.message));
+            tl.setResult(tl.TaskResult.Failed, tl.loc("MGAgentOperationOnAllVMsFailed", operation, exception.message));
+            this.deferred.reject(tl.loc("MGAgentOperationOnAllVMsFailed", operation, exception.message));
+        }
+    }
+
+    private setTaskResult(operation) {
+        if (this.failureCount + this.successCount == this.vmCount) {
+            if (this.failureCount > 0) {
+                this.log(tl.loc("MGAgentOperationOnAllVMsFailed", operation, ""));
+                this.deferred.reject(operation);
+            }
+            else {
+                this.log(tl.loc("MGAgentOperationOnAllVMsSucceeded", operation));
+                this.deferred.resolve(operation);
+            }
+        }
+    }
+
+    private givePostOperationCallBack(extensionName, vmName, operation) {
+        var postOperationCallBack = (error, result, request, response) => {
+            if (error) {
+                this.failureCount++;
+                this.errors += error.message;
+                this.errors += "\n";
+                this.log(tl.loc("OperationFailed", operation, extensionName, vmName));
+            } else {
+                this.successCount++;
+                this.log(tl.loc("OperationSucceeded", operation, extensionName, vmName));
+            }
+            this.setTaskResult(operation);
```

Comment:
When extension installation fails, the task fails. When extension deletion fails, the task does not fail. the method name 'setTaskResult' is not appropriate, since it only settles promises and does not set any task results. I have changed the name to '_settlePromise'.